presto-main/src/main/java/com/facebook/presto/execution/buffer/ArbitraryOutputBuffer.java

```diff
@@ -395,7 +409,7 @@ private synchronized ClientBuffer getBuffer(OutputBufferId id)
         return buffer;
     }
 
-    private synchronized Collection<ClientBuffer> safeGetBuffersSnapshot()
+    private synchronized List<ClientBuffer> safeGetBuffersSnapshot()
```

Comment:
I'm not sure why this is written this way. Why materialize this list on read every time? Why not materialize it only when the set of buffers changes?

Orthogonal to your set of changes though.

Reply:
I looked at that too, but shied away from it because of the on demand client buffer creation inside of “get”. Keeping the two structures in sync seemed brittle, but can definitely be done.

Reply:
Yeah. It wouldn't be difficult to make not brittle just because the buffers is an append only structure. A CAS loop on updating the materialized list would be enough.

Totally out of scope for this just seems like LHF.